libsmtutil/CHCSmtLib2Interface.cpp

```diff
@@ -64,50 +63,55 @@ void CHCSmtLib2Interface::registerRelation(Expression const& _expr)
 	if (!m_variables.count(_expr.name))
 	{
 		auto fSort = dynamic_pointer_cast<FunctionSort>(_expr.sort);
-		string domain = m_smtlib2->toSmtLibSort(fSort->domain);
+		string domain = toSmtLibSort(fSort->domain);
 		// Relations are predicates which have implicit codomain Bool.
 		m_variables.insert(_expr.name);
 		write(
-			"(declare-rel |" +
+			"(declare-fun |" +
 			_expr.name +
 			"| " +
 			domain +
-			")"
+			" Bool)"
 		);
 	}
 }
 
-void CHCSmtLib2Interface::addRule(Expression const& _expr, std::string const& _name)
+void CHCSmtLib2Interface::addRule(Expression const& _expr, std::string const& /*_name*/)
 {
 	write(
-		"(rule (! " +
+		"(assert\n(forall " + forall() + "\n" +
 		m_smtlib2->toSExpr(_expr) +
-		" :named " +
-		_name +
-		"))"
+		"))\n\n"
 	);
 }
 
 pair<CheckResult, CHCSolverInterface::CexGraph> CHCSmtLib2Interface::query(Expression const& _block)
 {
 	string accumulated{};
 	swap(m_accumulatedOutput, accumulated);
-	for (auto const& var: m_smtlib2->variables())
-		declareVariable(var.first, var.second);
+	solAssert(m_smtlib2, "");
+	writeHeader();
+	for (auto const& [type, decl]: m_smtlib2->userSorts())
+		write(decl);
```

Comment:
Maybe a range-v3 for only values might be better.

Reply:
Done